### PR TITLE
Additional REST and WebSocket logging

### PR DESCRIPTION
**Pull request by z64**
Ref #89, continues #113 

Logs:
- REST in/out, and rate limiting. Prints in/out bodies on DEBUG severity
- WS in/out
- Client messages identified by shard key

Todo:
- [x] ~~Probably token redaction. Unsure of where to place this, though.. suggestions? I imagine we'll have to sub inside of `Logger` but that doesn't sound fun since we don't know what client we're coming from. Will probably have to use some kind of token regex I guess.~~ After talking, users of discordcr should be fully aware of what the token is to not accidentally leak it somewhere.

### CODE REVIEW

**Comment by z64** [comment]
I'm thinking we should change `Client` to accept an `@logger`, with one that defaults to STDOUT instead of a single `LOGGER`. That would simplify things and I think ultimately be more flexible for the user.

File: src/discordcr/rest.cr [hunk 1]
```diff
@@ -34,6 +34,12 @@ module Discord
         mutexes[rate_limit_key].synchronize { }
         global_mutex.synchronize { }
 
+        if LOGGER.debug?
```
Comment (RX14) [inline src/discordcr/rest.cr]: bad. log the info in info AND the body in debug (if it's enabled).

```cr
LOGGER.info "[HTTP OUT] #{method} #{path} (#{body.try &.size || 0} bytes)"
LOGGER.debug "[HTTP OUT] BODY: #{body}" if LOGGER.debug?
```

File: src/discordcr/rest.cr [hunk 2]
```diff
@@ -52,8 +58,10 @@ module Discord
           end
 
           if response.headers["X-RateLimit-Global"]?
+            LOGGER.warn "Locking global mutex for #{retry_after}"
```
Comment (RX14) [inline src/discordcr/rest.cr]: How about just s/Locking global mutex/Pausing requests/

File: src/discordcr/rest.cr [hunk 3]
```diff
@@ -52,8 +58,10 @@ module Discord
...
             global_mutex.synchronize { sleep retry_after }
           else
+            LOGGER.warn "Locking #{rate_limit_key} mutex for #{retry_after} (#{response.status_code == 429 ? "too many requests" : "preemptively"})"
```
Comment (RX14) [inline src/discordcr/rest.cr]: ditto

File: src/discordcr/rest.cr [hunk 4]
```diff
@@ -65,6 +73,10 @@ module Discord
         end
       end
 
+      if response
+        LOGGER.debug "[HTTP IN] #{response.status_code} #{response.status_message} (#{response.body.size} bytes): #{response.body}"
```
Comment (RX14) [inline src/discordcr/rest.cr]: ditto on the `info` and `debug` split?

File: src/discordcr/websocket.cr [hunk 5]
```diff
@@ -23,6 +35,7 @@ module Discord
 
     def on_message(&handler : Packet ->)
       @websocket.on_message do |message|
+        LOGGER.debug "[WS IN] #{message}"
```
Comment (RX14) [inline src/discordcr/websocket.cr]: avoid throwing these strings away with `if LOGGER.debug?`

File: src/discordcr/websocket.cr [hunk 6]
```diff
@@ -32,7 +45,12 @@ module Discord
       @websocket.on_close(&handler)
     end
 
-    delegate run, close, send, to: @websocket
+    delegate run, close, to: @websocket
+
+    def send(message)
+      LOGGER.debug "[WS OUT] #{message}"
```
Comment (RX14) [inline src/discordcr/websocket.cr]: ditto

File: src/discordcr/rest.cr (hunk 2 above)
Comment (z64) [reply src/discordcr/rest.cr]: I think its important for the user to know they've actually hit the global rate limit and thus all requests will be affected, vs only a single route being locked.

**Comment by RX14** [comment]
need rebase

**Comment by z64** [comment]
@RX14 Rebased

File: src/discordcr/client.cr [hunk 7]
```diff
@@ -72,7 +72,9 @@ module Discord
                    @shard : Gateway::ShardKey? = nil,
                    @large_threshold : Int32 = 100,
                    @compress : Bool = false,
-                   @properties : Gateway::IdentifyProperties = DEFAULT_PROPERTIES)
+                   @properties : Gateway::IdentifyProperties = DEFAULT_PROPERTIES,
+                   @logger = Logger.new(STDOUT))
+      @logger.progname = "discordcr"
```
Comment (z64) [inline src/discordcr/client.cr]: Should I build the below `@client_name` into this? I pass the `@logger` to our `WebSocket` instance which won't have `shard` information. On the other hand, `WebSocket` only has debug level logging at the moment, which you probably wouldn't use anyways for a production/sharded application.

**Pull request merged**